### PR TITLE
Port to S4

### LINKED ISSUE #138: Port global-workflow to S4
The global-workflow should be ported to S4.

Branch: [DavidHuber-NOAA:port2s4](https://github.com/DavidHuber-NOAA/global-workflow/tree/port2s4)

Below is a checklist taken from #5 for porting to Orion which I will use as a template:

- [x]  Test: connection to GitHub
- [x]  Build: hpc-stack (personal area)
- [x]  Build: hpc-stack (official area)
- [x]  Install: manage_externals manic-v1.1.8 tag (personal area)
- [x]  Build: FV3GFS (w/WW3)
- [x]  Build: ProdGSI
- [x]  Build: GLDAS
- [x]  Build: UFS_UTILS
- [x]  Build: gfs_wafs
- [x]  Build: obsproc
- [x]  Build: chgres_recenter
- [x]  Build: chgres_recenter_ncio
- [x]  Build: fbwndgfs
- [x]  Build: fv3nc2nemsio
- [x]  Build: gaussian_sfcanl
- [x]  Build: gfs_bufr
- [x]  Build: regrid_nemsio
- [x]  Build: supvit
- [x]  Build: syndat_getjtbul
- [x]  Build: syndat_maksynrc
- [x]  Build: syndat_qctropcy
- [x]  Build: tave
- [x]  Build: tocsbufr
- [x]  Build: vint
- [x]  Test: fv3gfs_chgres.sh
- [x]  Establish: "glopara" space
- [x]  Establish: FIX_DIR under "glopara"
- [x]  Establish: BASE_GIT under "glopara"
- [x]  Establish: GDA under "glopara"
- [x]  Establish: obsproc_prep and obsproc_global under "glopara"
- [x]  Establish: Fit2Obs under "glopara"
- [x]  Establish: METplus
- [x]  Establish: ens_tracker under "glopara"
- [x]  Create: standard S4 environment
- [x]  Add: system paths to config.base.emc.dyn
- [x]  Add: system to machine-setup.sh
- [x]  Add: system to modulefiles/module_setup.sh[csh].inc
- [x]  Add: system to setup_expt.py
- [x]  Add: system to setup_expt_fcstonly.py
- [x]  Add: system to setup_workflow.py
- [x]  Add: system to setup_workflow_fcstonly.py
- [x]  Add: system to workflow_utils.py
- [x]  Review: S4 module files for final modules and hpc-stack
- [x]  Test: run setup_expt.py
- [x]  Test: run setup_expt_fcstonly.py
- [x]  Test: run setup_workflow.py
- [x]  Test: run setup_workflow_fcstonly.py
- [x]  Test: crons
- [x]  Test: ROCOTO
- [x]  Test: gdas[gfs]fcst job
- [x]  Test: gdas[gfs]post job
- [x]  Test: gdasvrfy job
- [x]  Test: gfsvrfy job
- [x]  Test: METplus jobs
- [x]  Test: gdas[gfs]arch job
- [x]  Test: complete free-forecast
- [x]  Test: gdas[gfs]prep job
- [x]  Test: gdaseobs job
- [x]  Test: gdas[gfs]anal job
- [x]  Test: gdas[gfs]analcalc job
- [x]  Test: gdasanaldiag job
- [x]  Test: gdasgldas job
- [x]  Test: gdasediag job
- [x]  Test: gdaseupd job
- [x]  Test: gdasecen job
- [x]  Test: gdasesfc job
- [x]  Test: gdasefcs job
- [x]  Test: gdasepos job
- [x]  Test: gdas[gfs]waveinit job
- [x]  Test: gdas[gfs]waveprep job
- [x]  Test: gdas[gfs]wavepostsbs job
- [x]  Test: GSI monitoring jobs
- [x]  Test: Fit2Obs
- [x]  Test: ens_tracker
- [x]  Test: C96C48 cycled
- [x]  Test: C192C96 cycled
- [x]  Test: C384C192 cycled
- [x]  Test: C96 ff
- [x]  Test: C192 ff
- [x]  Test: C384 ff
- [x]  Test: beta-testers
- [ ]  Pre-commit testing: cycled and ff on Hera
- [ ]  Pre-commit testing: cycled and ff on WCOSS-Cray
- [ ]  Pre-commit testing: cycled and ff on WCOSS-Dell
- [ ]  Perform review
- [x]  Supported in checkout: ufs-weather-model/fv3atm
- [x]  Supported in checkout: ProdGSI
- [x]  Supported in checkout: GLDAS
- [x]  Supported in checkout: EMC_post
- [x]  Supported in checkout: UFS_UTILS
- [x]  Supported in checkout: EMC_verif-global

### PR DESCRIPTION
**Pull request by DavidHuber-NOAA**
**Description**
Ports the global workflow to the S4 cluster.  Note that this does not include support for S2S experiments.  Additionally, S4 does not support C768 experiments.

A couple of special notes:

- S4 does not have access to rstprod data.  Among other things, this means that the nsstbufr and prepbufr.acft_profiles files must be created on the fly.  The way I accomplished this was by moving the `MAKE_NSSTBUFR` to and creating `MAKE_ACFTBUFR` in config.base.emc.dyn and setting them via setup_workflow.xml.  This seems like a bit of a kludge and I am open to suggestions on how else to address this.  Both options need to be set for the prep and analysis jobs.
- S4 _can_ run S2S+ experiments, but this requires significant, and convoluted, modifications to the configuration files.  Support for these are thus not enabled by default.  Instead, I have placed a set of configuration files in S4:/data/users/dhuber/save/s2s_configs. Users interested in performing these experiments should contact me to set it up.

**Type of change**

Fixes #138 

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)
- [X] This change requires a documentation update

*Suggested documentation updates*

- Under [global-workflow-components->data](https://github.com/NOAA-EMC/global-workflow/wiki/Global-Workflow-Components#data) add "S4: /data/prod/glopara/dump"
- [Under HPC Settings & Help](https://github.com/NOAA-EMC/global-workflow/wiki/HPC-Settings-&-Help)
   - [HPC Helpdesks](https://github.com/NOAA-EMC/global-workflow/wiki/HPC-Settings-&-Help#hpc-helpdesks) - add "S4: david.huber@noaa.gov"
   - [Restricted data access](https://github.com/NOAA-EMC/global-workflow/wiki/HPC-Settings-&-Help#restricted-data-access)  - add a note **"NOTE: Only non-restricted data is available on S4."**
   - Change the title of [HPC Settings & Help -> Stacksize on Hera & Orion](https://github.com/NOAA-EMC/global-workflow/wiki/HPC-Settings-&-Help#stacksize-on-hera--orion) to "Stacksize on Hera, Orion, & S4"
   - Add a new section titled "Optimizing the global workflow on S4" with the following text
      - "The S4 cluster is relatively small and so optimizations are recommended to improve cycled runtimes.  Please contact David Huber (david.huber@noaa.gov) if you are planning on running a cycled experiment on this system to obtain optimized configuration files."
   - [Version](https://github.com/NOAA-EMC/global-workflow/wiki/HPC-Settings-&-Help#version) - add the following row to the table:

| Machine | Default | Recommended |
| ---- | ---- | ---- |
| S4 | v1.8.3.1 | `module load git/2.30.0` |

  - Under [Clone and build global workflow](https://github.com/NOAA-EMC/global-workflow/wiki/Clone-and-build-global-workflow)
     - [Quick instructions](https://github.com/NOAA-EMC/global-workflow/wiki/Clone-and-build-global-workflow#quick-instructions) - Add `[s4]` to the `link_workflow.sh` command.
   - Under [Prepare initial conditions](https://github.com/NOAA-EMC/global-workflow/wiki/Prepare-initial-conditions)
      - [Manual generation, cold starts](https://github.com/NOAA-EMC/global-workflow/wiki/Prepare-initial-conditions#cold-starts) - Prepend the note "NOTE: Initial conditions **cannot** be generated on S4.  These must be generated on another supported platform then pushed to S4.  If you do not have access to a supported system or need assistance, please contact David Huber (david.huber@noaa.gov)."  References to dell and cray should be removed from this section, too.
   - Under [Run setup scripts to generate experiment](https://github.com/NOAA-EMC/global-workflow/wiki/Run-setup-scripts-to-generate-experiment) add the following command for python on S4 `module load miniconda/3.8-s4` and for rocoto `module load rocoto/1.3.4`.
   - Under [Start your run -> set up your experiment cron](https://github.com/NOAA-EMC/global-workflow/wiki/Start-your-run#set-up-your-experiment-cron), remove subtitle "HPCs with access to directly edit your crontab files (WCOSS-Cray, Hera, Jet, Orion)" and the subsection "HPCs without access to directly edit your crontab files (WCOSS-Dell)"

**How Has This Been Tested?**

  
**Checklist**

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [X] Any dependent changes have been merged and published

### CODE REVIEW

**Review by aerorahul — COMMENTED** [review]
@DavidHuber-NOAA 
I noticed several places where a check is placed to test if the resolution on S4 exceeds a threshold.
I have opened a PR #1026 that does this check at a much higher level and would suggest removing those checks from this PR for S4.

**Comment by DavidHuber-NOAA** [comment]
@aerorahul OK, I have updated the port to exclude the resolution restrictions.  Thanks for setting up a separate PR for that.

**Comment by DavidHuber-NOAA** [comment]
@aerorahul @KateFriedman-NOAA @WalterKolczynski-NOAA All of the linter/shellcheck warnings and errors have been addressed and this PR is now ready for your reviews.  I will note that I took a few liberties when defining defaults in S4.env to resolve shellcheck warnings.

File: env/S4.env
Comment (WalterKolczynski-NOAA) [inline env/S4.env]: We've converted all workflow scripts to bash, so this should be as well.

File: env/S4.env
Comment (WalterKolczynski-NOAA) [inline env/S4.env]: Please remove dead code

File: jobs/JGDAS_ENKF_DIAG
Comment (WalterKolczynski-NOAA) [inline jobs/JGDAS_ENKF_DIAG]: "MAKE" is already a direct verb; using "DO" as well is probably unnecessary.

File: jobs/JGDAS_ENKF_SELECT_OBS
Comment (WalterKolczynski-NOAA) [inline jobs/JGDAS_ENKF_SELECT_OBS]: Same

File: jobs/JGLOBAL_ATMOS_ANALYSIS
Comment (WalterKolczynski-NOAA) [inline jobs/JGLOBAL_ATMOS_ANALYSIS]: Same

File: parm/config/config.fv3
Comment (WalterKolczynski-NOAA) [inline parm/config/config.fv3]: Still don't like this carve-out for a machine few are likely to use. @aerorahul?

File: sorc/machine-setup.sh
Comment (WalterKolczynski-NOAA) [inline sorc/machine-setup.sh]: Move this up a line inside the if-block like the command it is undoing.

File: workflow/hosts/hera.yaml
Comment (WalterKolczynski-NOAA) [inline workflow/hosts/hera.yaml]: As before.

File: workflow/hosts/orion.yaml
Comment (WalterKolczynski-NOAA) [inline workflow/hosts/orion.yaml]: Same

File: workflow/hosts/hera.yaml
Comment (WalterKolczynski-NOAA) [inline workflow/hosts/hera.yaml:15]: Were these tested on Hera and Orion to confirm?

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

**Comment by KateFriedman-NOAA** [comment]
@DavidHuber-NOAA Was your branch tested on either Hera or Orion to ensure it didn't break functionality on the other R&D platforms?

File: parm/config/config.aero
Comment (KateFriedman-NOAA) [inline parm/config/config.aero:17]: Are you keeping this sync'd with the Hera copy?

File: parm/config/config.fv3
Comment (KateFriedman-NOAA) [inline parm/config/config.fv3:52]: Have these values been tested on one of the other R&Ds to check they work there as well? These lines are used on all platforms.

File: parm/config/config.fv3
Comment (KateFriedman-NOAA) [reply parm/config/config.fv3]: I agree with @WalterKolczynski-NOAA . @DavidHuber-NOAA I'm guessing a 6x4 layout for C96 gdasfcst fails on S4? Is 6x4 for gfsfcst ok?

Comment (KateFriedman-NOAA) [reply parm/config/config.fv3]: Similar questions about the other resolution S4 if-blocks.

File: env/S4.env
Comment (DavidHuber-NOAA) [reply env/S4.env]: Done.

File: env/S4.env
Comment (DavidHuber-NOAA) [reply env/S4.env]: Done.

File: jobs/JGDAS_ENKF_DIAG
Comment (DavidHuber-NOAA) [reply jobs/JGDAS_ENKF_DIAG]: Changed this to MAKE_*.  I also changed DO_MAKEPREPBUFR to MAKE_PREPBUFR to match.

File: parm/config/config.fv3
Comment (DavidHuber-NOAA) [reply parm/config/config.fv3]: @KateFriedman-NOAA A 6x4 layout for C96 runs just fine on S4 individually, but makes for a significantly slower cycle time.  Each user is restricted to 40 nodes at a given time.  A 4x4 job uses 4 nodes while a 6x4 uses 6 (10 jobs at a time vs 6) and the gain in runtime does not come close to making up for the extra time required to run 10 jobs 4 time vs 6 jobs 7 times.

File: sorc/machine-setup.sh
Comment (DavidHuber-NOAA) [reply sorc/machine-setup.sh]: This is a feature of shellcheck directives.  The next line of code cannot be a block ending or partial statement.  See [this note](https://www.shellcheck.net/wiki/SC1123).  I could move the directive before the if block to match tiers, but that may be confusing as to what it is there for.

File: parm/config/config.fv3
Comment (aerorahul) [reply parm/config/config.fv3]: IMO, the resource allocation needs to be thought over again.  We have machines ranging from S4 to WCOSS2 which have significant differences in the node/core footprint.
I don't want to hold this PR up, but I would suggest until we get that piece resolved, we may have to sacrifice efficiency on S4. 
Happy to hear your thoughts.

File: workflow/hosts/hera.yaml
Comment (DavidHuber-NOAA) [reply workflow/hosts/hera.yaml]: Done.

File: workflow/hosts/orion.yaml
Comment (DavidHuber-NOAA) [reply workflow/hosts/orion.yaml]: Done.

File: workflow/hosts/hera.yaml
Comment (DavidHuber-NOAA) [reply workflow/hosts/hera.yaml:15]: I tested this on Hera.  I should also test on Orion.

**Comment by DavidHuber-NOAA** [comment]
> @DavidHuber-NOAA Was your branch tested on either Hera or Orion to ensure it didn't break functionality on the other R&D platforms?

Yes, I tested this on Hera.

File: parm/config/config.aero
Comment (DavidHuber-NOAA) [reply parm/config/config.aero:17]: I sync'd it two weeks ago, but haven't been keeping up since then.  I will add it to the nightly rsync, though.  Thanks.

Comment (KateFriedman-NOAA) [reply parm/config/config.aero:17]: Cool...this set doesn't update too much (same with fix) so maybe a weekly check is sufficient (or after an announcement) but I leave that up to you. Recommend using the `--checksum` flag with rsync from time to time to catch any diffs that aren't caught by the usual checks rsync does on metadata and filesize. Using that flag significantly increases rsync time though so not needed all the time.

File: parm/config/config.fv3
Comment (DavidHuber-NOAA) [reply parm/config/config.fv3:52]: @KateFriedman-NOAA Yes, thanks for reminding me.  During comparison testing on Hera, these jobs failed for running out of memory.  You had suggested in an email that I include this fix in this port.  I can separate it to its own PR if desired, though.

Comment (KateFriedman-NOAA) [reply parm/config/config.fv3:52]: If changing the C48 values also helps on Hera then I am happy to have them stay in this PR. Thanks for reminding me we discussed that earlier! @aerorahul thoughts?

@DavidHuber-NOAA Do you happen to know if this works and/or helps on Orion as well? Orion has more memory than Hera so it may not matter in terms of memory but would want to know it doesn't fail. Thanks!

Comment (DavidHuber-NOAA) [reply parm/config/config.fv3:52]: @KateFriedman-NOAA I believe I tested this on Orion as well, but it was some time ago and I can't find the log files.  That said, I'm going to start a few cycled tests on Hera and Orion to run over the weekend and I will include this.

File: parm/config/config.fv3
Comment (DavidHuber-NOAA) [reply parm/config/config.fv3]: @aerorahul I can understand that and I can pull these changes.  However, I would like to note that for C384/C192 experiments, it is a significant cycle time difference (~2.25 hours/cycle) and does take up a lot of extra resources.

That said, perhaps a compromise would be to ask users on the wiki to contact me to help optimize their experiments when running on S4.  I can then keep these configuration files on S4 for reference.

File: parm/config/config.fv3
Comment (DavidHuber-NOAA) [reply parm/config/config.fv3:52]: @KateFriedman-NOAA The C48 ensemble forecasts ran correctly for the first half cycle on Hera.  I'm still waiting on Orion.

A side note, the gdasanal job failed on the first full cycle (2020080100).  For this cycle, the script is expecting that the diag files in the 2020073118 radstat file are netCDF4 files.  However, they are instead the older binary format.  I'm going to investigate this a little more and will write a bug report shortly.

Comment (DavidHuber-NOAA) [reply parm/config/config.fv3:52]: @KateFriedman-NOAA Ah, I now see the `binary_diag` option in config.base.  I will give it another go with that `=".true."`.

File: parm/config/config.fv3
Comment (DavidHuber-NOAA) [reply parm/config/config.fv3]: I've removed most of the optimizations from config.fv3 and config.resources.  I left just the resource modifications necessary for the analysis jobs (gdasanal, gfsanal, and gdaseobs).  This was discussed by email with @CoryMartin-NOAA and [summarized here for Jet](https://github.com/NOAA-EMC/global-workflow/issues/357#issuecomment-909483710).  The basics of that conversation are that if the number of cores allocated differs from the number of cores needed by these jobs, then some data may not be assimilated and will result in apparent spatial data gaps.  This is a result of S4 having 32 cores/node vs Hera and Orion's 40.

Below is a plot of the ATMS data points when using `npe_eobs=40` and `npe_node_eobs=16`.

Screen Shot 2021-05-27 at 9 22 30 AM

And here is the same plot when using `npe_eobs=40` and `npe_node_eobs=10`:

Screen Shot 2021-05-27 at 2 59 29 PM

File: workflow/hosts/hera.yaml
Comment (DavidHuber-NOAA) [reply workflow/hosts/hera.yaml:15]: Confirmed that this also works on Orion.

File: parm/config/config.fv3
Comment (DavidHuber-NOAA) [reply parm/config/config.fv3:52]: This fixed the C96 analysis jobs.  The gdaspostanl and gfspostanl jobs are both failing, as reported in #1006.

The gdaseupd job is failing when `binary_diag=".true."`.  I believe this is a separate issue and will report it after checking if it exists at C96 ensemble resolution.  The problem seems to stem from the output from enkf.x not being in the expected binary format.  When run with newer diag files and `netcdf_diag=.true."` at C48 resolution, the job runs correctly.

**Comment by WalterKolczynski-NOAA** [comment]
Once conflicts are resolved, I think this is ready

**Comment by DavidHuber-NOAA** [comment]
This port needs the gfs-utils repository to be ported.  I'm working on that now (NOAA-EMC/gfs-utils#5).

**Comment by DavidHuber-NOAA** [comment]
I believe this branch is now ready and that all comments have been addressed.

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Review by KateFriedman-NOAA — APPROVED** [review]

**Comment by KateFriedman-NOAA** [comment]
@DavidHuber-NOAA FYI, another PR about to go in will change the obsproc/prepobs packages, as well as how prepobs is used (now a module). You will need to install both packages on S4 and update the `module_base.s4.lua` to load the prepobs module.

See https://github.com/NOAA-EMC/global-workflow/pull/1049.

The new packages:

obsproc:
Orion install: /work/noaa/global/glopara/git/obsproc/v1.0.2
source: https://github.com/NOAA-EMC/obsproc.git
tag: OT.obsproc.v1.0.2_20220628

prepobs:
Orion install: /work/noaa/global/glopara/git/prepobs/v1.0.1
source: https://github.com/NOAA-EMC/prepobs.git
tag: OT.prepobs.v1.0.1_20220628

Note, they don't currently support S4 so you'll need to make modifications for them to build on S4.

To install (once updated for S4):

obsproc:
```
cd build-obsproc
INSTALL_TARGET=s4 ./build.sh
```

prepobs:
```
INSTALL_TARGET=s4 INSTALL_PREFIX=PATH_TO_TOP_OF_INSTALL ./build.sh
```

A new version of obsproc (v1.1.0) will be coming in with GFSv16.3 next month and `develop` will be updating to it soon after. Look for that update so you can install that version on S4 as well. The prepobs package will be staying at v1.0.1 with GFSv16.3.

**Comment by DavidHuber-NOAA** [comment]
@KateFriedman-NOAA Thanks for the heads up.  I wasn't sure what the order would be for this, so I ported those packages and built them on S4 yesterday, so they should be good to go.  I will run a test cycle today just to be sure and update `module_base.s4.lua`.

**Comment by KateFriedman-NOAA** [comment]
@DavidHuber-NOAA Ok cool! Final review of the obsproc PR is happening now and I hope to merge it today.

**Comment by KateFriedman-NOAA** [comment]
@DavidHuber-NOAA FYI PR #1049 (obsproc update) has just been merged into `develop`.

**Comment by DavidHuber-NOAA** [comment]
@KateFriedman-NOAA A cycled test completed successfully with the updated prepobs and obsproc packages on S4.

**Comment by KateFriedman-NOAA** [comment]
> A cycled test completed successfully with the updated prepobs and obsproc packages on S4.

Excellent, thanks for confirming @DavidHuber-NOAA ! My approval of this PR stands. :)

**Comment by WalterKolczynski-NOAA** [comment]
@DavidHuber-NOAA Please make the needed additions to the wiki when you get the chance.

**Comment by DavidHuber-NOAA** [comment]
@WalterKolczynski-NOAA I do not have edit privileges for the Wiki.  However, I did put my suggested edits in the description of this PR.  Would someone be able to add those in for me or give me temporary privileges?

**Comment by DavidHuber-NOAA** [comment]
@KateFriedman-NOAA @WalterKolczynski-NOAA Would one of you be able to give me permissions to edit the wiki so I can add info on S4?  If that's not possible, I have suggested edits at the top of this PR -- would you be able to add this for me?  Thanks!

**Comment by KateFriedman-NOAA** [comment]
@DavidHuber-NOAA I have added your S4 updates into the global-workflow GitHub wiki. I cleaned some of the dell/cray references but not all since some (ICs section) will take more thought on another day.

**Comment by DavidHuber-NOAA** [comment]
@KateFriedman-NOAA Thank you!

**Pull request merged**
